### PR TITLE
Fixes #6043 - Reimplement UnixSocket support based on Java 16.

### LINKED ISSUE #6043: Reimplement UnixSocket support based on Java 16
**Jetty version**
10.0.x

**Description**
Java 16 provides an implementation of Unix Sockets ([JEP 380](https://openjdk.java.net/jeps/380)).

We should provide an implementation based on that rather than the more unstable JNR.

### PR DESCRIPTION
**Pull request by sbordet**
* Introduced new module "jetty-server-unixdomain".
It uses reflection to access the Java 16 Unix-Domain classes to keep compatibility with the other modules and the build.
* Added Jetty module with only HTTP/1.1 support for now (requires review of the modules to reuse them with various connectors).
* Updated documentation to mention UnixDomainServerConnector.
* Updated client libraries to support Unix-Domain.
* Updated PROXY protocol implementation to support Unix-Domain.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

### CODE REVIEW

File: jetty-unixdomain-server/src/main/config/etc/jetty-unixdomain.xml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  <Call name="addConnector">
+    <Arg>
+      <New id="unixDomainConnector" class="org.eclipse.jetty.unixdomain.server.UnixDomainServerConnector">
+        <Arg><Ref refid="Server" /></Arg>
+        <Arg type="int"><Property name="jetty.unixdomain.acceptors" default="1"/></Arg>
+        <Arg type="int"><Property name="jetty.unixdomain.selectors" default="-1"/></Arg>
+        <Arg>
+          <Array type="org.eclipse.jetty.server.ConnectionFactory">
+            <Item>
+              <New class="org.eclipse.jetty.server.HttpConnectionFactory">
```
Comment (gregw) [inline jetty-unixdomain-server/src/main/config/etc/jetty-unixdomain.xml]: Is `HTTP` the only protocol that we will deliver over UnixDomain sockets?
Can we do `SSL`, `Proxy`, `HTTP/2` ? 

At the very least, this file should be named `jetty-unixdomain-http.xml`

File: jetty-unixdomain-server/src/main/config/modules/unixdomain.mod [hunk 2]
```diff
@@ -0,0 +1,36 @@
+[description]
+Enables support for Java 16 Unix-Domain server sockets.
```
Comment (gregw) [inline jetty-unixdomain-server/src/main/config/modules/unixdomain.mod]:
```suggestion
Enables support for HTTP over Java 16 Unix-Domain server sockets.
```
rename module to `unixdomain-http` or `http-unixdomain`

**Review by gregw — CHANGES_REQUESTED** [review]
I'm concerned with hard coding (in config xml) that unix domain is only for raw HTTP.   I would expect that other protocols may be carried (eg PROXY from haproxy), so we should set up to be a bit more flexible.

File: pom.xml [hunk 3]
```diff
@@ -150,6 +150,7 @@
     <module>jetty-bom</module>
     <module>documentation</module>
     <module>jetty-keystore</module>
+    <module>jetty-unixdomain-server</module>
```
Comment (olamy) [inline pom.xml:153]: this module can be included only within a jdk16 maven profile.

**Review by olamy — CHANGES_REQUESTED** [review]
the profile <id>eclipse-release</id> must be updated as well to enforce usage of jdk16 when releasing

File: pom.xml (hunk 3 above)
Comment (sbordet) [reply pom.xml:153]: It does not require Java 16 to _build_, only to be used, so there is no need for profiles, not in the build nor in the release.

File: jetty-unixdomain-server/src/main/config/etc/jetty-unixdomain.xml (hunk 1 above)
Comment (sbordet) [reply jetty-unixdomain-server/src/main/config/etc/jetty-unixdomain.xml]: @gregw I agree, but the current status for the other implementation is that we have 7 modules for all possible permutations.

I wanted to explore the possibility to separate the "carrier", i.e. IP vs Unix-Domain, from the protocols you attach to it.
Something along the lines of `--add-modules=http2` would implicitly use IP (i.e. `ServerConnector`), while `--add-modules=unix-domain,http2` would use Unix-Domain (i.e. `UnixDomainServerConnector`).

I thought that this could be worked out in a different PR.

File: pom.xml (hunk 3 above)
Comment (olamy) [reply pom.xml:153]: in this case you need to remove this https://github.com/eclipse/jetty.project/pull/6522/files#diff-723a454092742efde002c05bbdc642cd7c1a3a4b1d069bffe9ab63fdb1c10bfbR24
```
        <configuration>
          <source>16</source>
          <target>16</target>
          <release>16</release>
        </configuration>
```
because this doesn't build with jdk11.
```
release version 16 not supported
```

Comment (sbordet) [reply pom.xml:153]: Ah, good catch!

File: jetty-unixdomain-server/src/main/config/etc/jetty-unixdomain.xml (hunk 1 above)
Comment (gregw) [reply jetty-unixdomain-server/src/main/config/etc/jetty-unixdomain.xml]: Working out the full topology can be punted to a different PR, but you need to at least start with the right name.    This should be `unixdomain-http` with initially can be complete in itself, but in future might depend on a `unix-domain` module.      If you start with the non protocol module name, then we have less flexibility in future

File: documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/http/HTTPServerDocs.java [hunk 4]
```diff
@@ -165,12 +167,13 @@ public void configureConnector() throws Exception
         // Create a ServerConnector instance.
         ServerConnector connector = new ServerConnector(server, 1, 1, new HttpConnectionFactory());
 
-        // Configure TCP parameters.
+        // Configure IP parameters.
 
-        // The TCP port to listen to.
+        // The IP port to listen to.
```
Comment (lorban) [inline documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/http/HTTPServerDocs.java]: The port still relates to `TCP`.

File: documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/http/HTTPServerDocs.java [hunk 5]
```diff
@@ -179,6 +182,33 @@ public void configureConnector() throws Exception
...
+        Server server = new Server();
+
+        // The number of acceptor threads.
+        int acceptors = 1;
+
+        // The number of selectors.
+        int selectors = 1;
+
+        // Create a ServerConnector instance.
+        UnixDomainServerConnector connector = new UnixDomainServerConnector(server, 1, 1, new HttpConnectionFactory());
```
Comment (lorban) [inline documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/http/HTTPServerDocs.java]: Aren't you supposed to use the `acceptors` and `selectors` variables here?

File: jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java [hunk 6]
```diff
@@ -549,24 +549,25 @@ protected boolean removeDestination(HttpDestination destination)
...
     {
         HttpDestination destination = (HttpDestination)resolveDestination(request);
         destination.send(request, listeners);
     }
 
-    protected void newConnection(final HttpDestination destination, final Promise<Connection> promise)
+    protected void newConnection(HttpDestination destination, Promise<Connection> promise)
     {
+        // Multiple threads may access the map, especially with DEBUG logging enabled.
+        Map<String, Object> context = new ConcurrentHashMap<>();
```
Comment (lorban) [inline jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java:561]: Is this a read-only map or a modifiable one? In the former case, an unmodifiable `HashMap` would be preferable.

File: jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpClientTransportOverHTTP.java [hunk 7]
```diff
@@ -26,22 +26,24 @@
...
+    private static final Logger LOG = LoggerFactory.getLogger(HttpClientTransportOverHTTP.class);
 
     private final ClientConnectionFactory factory = new HttpClientConnectionFactory();
     private int headerCacheSize = 1024;
     private boolean headerCacheCaseSensitive;
 
     public HttpClientTransportOverHTTP()
     {
-        this(Math.max(1, ProcessorUtils.availableProcessors() / 2));
+        this(1);
```
Comment (lorban) [inline jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpClientTransportOverHTTP.java]: Are you sure you want to change this behavior in this PR?

File: jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java [hunk 8]
```diff
@@ -221,20 +238,16 @@ public void connect(SocketAddress address, Map<String, Object> context)
...
-                    LOG.debug("Binding to {} to connect to {}{}", bindAddress, address, (reuseAddress ? " reusing address" : ""));
-                channel.setOption(StandardSocketOptions.SO_REUSEADDR, reuseAddress);
-                channel.bind(bindAddress);
-            }
+                bind(channel, bindAddress);
             configure(channel);
 
             boolean connected = true;
-            boolean blocking = isConnectBlocking();
+            boolean blocking = isConnectBlocking() && address instanceof InetSocketAddress;
```
Comment (lorban) [inline jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java:255]: Why is `isConnectBlocking` ignored for non `InetSocketAddress` instances?

File: jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java [hunk 9]
```diff
@@ -288,9 +301,34 @@ public void accept(SocketChannel channel, Map<String, Object> context)
...
+        {
+            boolean reuseAddress = getReuseAddress();
+            if (LOG.isDebugEnabled())
+                LOG.debug("Binding to {} reusing address {}", bindAddress, reuseAddress);
+            channel.setOption(StandardSocketOptions.SO_REUSEADDR, reuseAddress);
+            channel.bind(bindAddress);
+        }
+        catch (Throwable x)
+        {
+            if (LOG.isDebugEnabled())
```
Comment (lorban) [inline jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java:321]: I'm not certain silently ignoring the bind address when bind fails is a sensible thing to do. Why do you want to change the existing behavior of throwing when bind fails?

File: jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java [hunk 10]
```diff
@@ -288,9 +301,34 @@ public void accept(SocketChannel channel, Map<String, Object> context)
...
+                LOG.debug("Could not bind {}", channel);
+        }
+    }
+
     protected void configure(SocketChannel channel) throws IOException
     {
-        channel.socket().setTcpNoDelay(true);
+        try
+        {
+            channel.setOption(StandardSocketOptions.TCP_NODELAY, true);
```
Comment (lorban) [inline jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java:330]: Why forcing this TCP option by default?

File: jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java [hunk 11]
```diff
@@ -351,4 +389,78 @@ protected void connectionFailed(SelectableChannel channel, Throwable failure, Ob
...
+             * derived from the input socket address.</p>
+             * <p>The input socket address represents the destination socket address to
+             * connect to, as it is typically specified by a URI authority, for example
+             * {@code localhost:8080} if the URI is {@code http://localhost:8080/path}.</p>
+             * <p>However, the returned socket address may be different as the implementation
+             * may use a Unix-Domain socket address to physically connect to the virtual
+             * destination socket address given as input.</p>
+             * <p>The return type is a pair/record holding the socket channel and the
+             * socket address, with the socket channel not yet connected.
+             * The implementation of this methos must not call
```
Comment (lorban) [inline jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java]: Typo: implementation of this **methos** -> **method**

File: jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java [hunk 12]
```diff
@@ -351,4 +389,78 @@ protected void connectionFailed(SelectableChannel channel, Throwable failure, Ob
...
+
+        public SocketAddress getSocketAddress()
+        {
+            return address;
+        }
+
+        /**
+         * <p>A factory for {@link SocketChannelWithAddress} instances.</p>
+         */
+        public interface Factory
```
Comment (lorban) [inline jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java]: This mechanism won't work for QUIC as it works with `DatagramChannel` instead of `SocketChannel`.

The `SocketChannelWithAddress` class could be modified to hold a `SelectableChannel` instead of `SocketChannel` but that would not solve the API problem that `ClientConnector` already exposes `SocketChannel` via `configure(SocketChannel channel)` for instance.

This class is sitting between OSI Layer 4 (TCP) and Layer 5 (Sockets, both TCP and Unix Domain), or maybe even overlapping both so I'm not sure I would expose it as an API. Something doesn't look right about it.

File: jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java [hunk 13]
```diff
@@ -351,4 +389,78 @@ protected void connectionFailed(SelectableChannel channel, Throwable failure, Ob
...
+        {
+            public static Factory forUnixDomain(Path path)
+            {
+                return (address, context) ->
+                {
+                    try
+                    {
+                        ProtocolFamily family = Enum.valueOf(StandardProtocolFamily.class, "UNIX");
+                        Class<?> channelClass = Class.forName("java.nio.channels.SocketChannel");
+                        SocketChannel socketChannel = (SocketChannel)channelClass.getMethod("open", ProtocolFamily.class).invoke(null, family);
```
Comment (lorban) [inline jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java]: How about `SocketChannel.class.getMethod("open", ProtocolFamily.class).invoke(null, family);`?

File: jetty-server/src/main/java/org/eclipse/jetty/server/ProxyConnectionFactory.java [hunk 14]
```diff
@@ -957,4 +970,47 @@ public void write(Callback callback, ByteBuffer... buffers) throws WritePendingE
...
+                    return (SocketAddress)unixDomainSocketAddress.getMethod("of", String.class).invoke(null, path);
+                return null;
+            }
+            catch (Throwable ignored)
+            {
+                return null;
+            }
+        }
+
+        private static String toPath(byte[] bytes)
```
Comment (lorban) [inline jetty-server/src/main/java/org/eclipse/jetty/server/ProxyConnectionFactory.java:1004]: This method looks unused.

**Review by lorban — CHANGES_REQUESTED** [review]

File: jetty-unixdomain-server/src/main/config/etc/jetty-unixdomain.xml (hunk 1 above)
Comment (sbordet) [reply jetty-unixdomain-server/src/main/config/etc/jetty-unixdomain.xml]: Done.

File: jetty-unixdomain-server/src/main/config/modules/unixdomain.mod (hunk 2 above)
Comment (sbordet) [reply jetty-unixdomain-server/src/main/config/modules/unixdomain.mod]: Done.

File: documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/http/HTTPServerDocs.java (hunk 5 above)
Comment (sbordet) [reply documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/http/HTTPServerDocs.java]: Good catch!

File: jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java (hunk 6 above)
Comment (sbordet) [reply jetty-client/src/main/java/org/eclipse/jetty/client/HttpClient.java:561]: It is a modifiable one, as different `ClientConnectionFactory` may enrich it with additional entries.

File: jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpClientTransportOverHTTP.java (hunk 7 above)
Comment (sbordet) [reply jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpClientTransportOverHTTP.java]: Good point, reverted.

File: jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java (hunk 8 above)
Comment (sbordet) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java:255]: Because in the code below I cannot call `channel.socket()` if it is a Unix-Domain socket address (it throws).
I would do `channel instanceof SocketChannel`, but that is true for both network and Unix-Domain, despite the fact that calling `socket()` on a `SocketChannel` may throw if it is Unix-Domain. 🤷🏼‍♂️

File: jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java (hunk 9 above)
Comment (sbordet) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java:321]: Again because for Unix-Domain, binding makes no sense, so if `ClientConnector` is misconfigured, we ignore the binding attempt.

File: jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java (hunk 10 above)
Comment (sbordet) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java:330]: You are right that this should be configurable, but it's covered by #6372.

File: jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java (hunk 13 above)
Comment (sbordet) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java]: Much better indeed.

File: jetty-server/src/main/java/org/eclipse/jetty/server/ProxyConnectionFactory.java (hunk 14 above)
Comment (sbordet) [reply jetty-server/src/main/java/org/eclipse/jetty/server/ProxyConnectionFactory.java:1004]: It's used at lines 614 and 616.

File: documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/http/HTTPServerDocs.java [hunk 15]
```diff
@@ -179,6 +182,33 @@ public void configureConnector() throws Exception
...
+
+        // Create a ServerConnector instance.
+        UnixDomainServerConnector connector = new UnixDomainServerConnector(server, acceptors, selectors, new HttpConnectionFactory());
+
+        // Configure Unix-Domain parameters.
+
+        // The Unix-Domain path to listen to.
+        connector.setUnixDomainPath(Path.of("/tmp/jetty.sock"));
+
+        // The TCP accept queue size.
```
Comment (lorban) [inline documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/server/http/HTTPServerDocs.java:204]: This is unrelated to TCP.

File: jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java (hunk 12 above)
Comment (joakime) [reply jetty-io/src/main/java/org/eclipse/jetty/io/ClientConnector.java]: This should probably be a new issue.

**Comment by sbordet** [comment]
@lorban I made the mechanism to create a Unix-Domain `ClientConnector` more private, as discussed.

I also updated `FastCGIProxyServlet` to use Unix-Domain when configured so.

**Review by lorban — DISMISSED** [review]

**Comment by sbordet** [comment]
@gregw @olamy nudge

**Review by gregw — DISMISSED** [review]
better in that out at this stage!

File: jetty-fcgi/fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/proxy/FastCGIProxyServletTest.java [hunk 16]
```diff
@@ -216,4 +230,34 @@ public void handle(String target, org.eclipse.jetty.server.Request baseRequest,
...
     }
+
+    @Test
+    @EnabledForJreRange(min = JRE.JAVA_16)
+    public void testUnixDomain() throws Exception
+    {
+        int maxUnixDomainPathLength = 108;
+        Path path = Files.createTempFile("unix", ".sock");
+        if (path.normalize().toAbsolutePath().toString().length() > maxUnixDomainPathLength)
+            path = Files.createTempFile(Path.of("/tmp"), "unix", ".sock");
```
Comment (olamy) [inline jetty-fcgi/fcgi-server/src/test/java/org/eclipse/jetty/fcgi/server/proxy/FastCGIProxyServletTest.java:241]: can be a problem on windows or osx? this test should not run on windows `@DisabledOnOs(OS.WINDOWS)`?

File: jetty-unixdomain-server/src/test/java/org/eclipse/jetty/unixdomain/server/UnixDomainTest.java [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+import org.junit.jupiter.api.condition.JRE;
+
+import static org.eclipse.jetty.client.ProxyProtocolClientConnectionFactory.V1;
+import static org.eclipse.jetty.client.ProxyProtocolClientConnectionFactory.V2;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
```
Comment (olamy) [inline jetty-unixdomain-server/src/test/java/org/eclipse/jetty/unixdomain/server/UnixDomainTest.java:57]: maybe need `@DisabledOnOs(OS.WINDOWS)`

File: tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java [hunk 18]
```diff
@@ -906,4 +908,37 @@ public void testDefaultLoggingProviderNotActiveWhenExplicitProviderIsPresent() t
             assertFalse(Files.exists(jettyBase.resolve("resources/jetty-logging.properties")));
         }
     }
+
+    @Test
+    @EnabledForJreRange(min = JRE.JAVA_16)
```
Comment (olamy) [inline tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java:903]: `@DisabledOnOs(OS.WINDOWS)`

**Review by olamy — CHANGES_REQUESTED** [review]
might need to disable some tests for windows OS

**Review by joakime — APPROVED** [review]
Tested on Windows 10 laptop, works as expected.

**Comment by sbordet** [comment]
@olamy this PR now runs fine on Windows too.
There is a new profile to set the Unix-Domain directory for Windows.
Property `unix.socket.tmp` has been removed in favor of `jetty.unixdomain.dir`.

**Review by olamy — APPROVED** [review]
